src/packet_analysis/protocol/ip/IPBasedAnalyzer.h

```diff
@@ -0,0 +1,182 @@
+
+protected:
+
+	IPBasedAnalyzer(const char* name);
+
+	/**
+	 * Entry point for child classes to call to do the actual heavy lifting for
+	 * processing a packet and extracting a connection out of it.
+	 *
+	 * @param conn_id The connection ID generated by the child class.
+	 * @param pkt The packet being processed.
+	 * @param remaining The number of bytes remaining to be processed in the packet.
+	 */
+	void ProcessConnection(const ConnTuple& conn_id, Packet* pkt, size_t remaining);
+
+	/**
+	 * Verifies that there is enough data in the packet to process the header
+	 * length requested.
+	 *
+	 * @param min_hdr_len The minimum data in bytes that needs to exist.
+	 * @param remaining The remaining number of bytes in the packet reported by
+	 * previous analyzer.
+	 * @param packet The packet being processed. This will be used to pull out the
+	 * number of bytes the IP header says we have remaining.
+	 */
+	bool CheckHeaderTrunc(size_t min_hdr_len, size_t remaining, Packet* packet);
+
+	/**
+	 * Upon seeing the first packet of a connection, checks whether we want
+	 * to analyze it (e.g. we may not want to look at partial connections)
+	 * and, if yes, whether we should flip the roles of originator and
+	 * responder based on known ports and such.
+	 *
+	 * @param src_port The source port of the connection.
+	 * @param dst_port The destination port of the connection.
+	 * @param data The payload data for the packet being processed.
+	 * @param flip_roles Return value if the roles should be flipped.
+	 * @return True if the connection is wanted. False otherwise.
+	 */
+	virtual bool WantConnection(uint16_t src_port, uint16_t dst_port,
+	                            const u_char* data, bool& flip_roles) const
+		{
+		flip_roles = false;
+		return true;
+		}
+
+	/**
+	 * Returns the port mask for an analyzer. Used by IsLikelyServerPort().
+	 */
+	virtual uint32_t GetServerPortMask() const = 0;
+
+	/**
+	 * Returns the transport protocol. Used by NewConn().
+	 */
+	virtual TransportProto GetTransportProto() const = 0;
+
+	/**
+	 * Returns true if the port corresponds to an application for which there
+	 * is a Zeek analyzer (even if it might not be used by the present policy
+	 * script) or if it's generally a likely server port.
+	 *
+	 * @param port The port number to check, in host order.
+	 */
+	bool IsLikelyServerPort(uint32_t port) const;
+
+	/**
+	 * Continues process of packet after the connection has been inserted into the
+	 * session manager. This should be implemented by all child classes.
+	 *
+	 * @param conn The connection currently being processed.
+	 * @param t The timestamp for the current packet.
+	 * @param is_orig Flag denoting whether this packet is from the originator of
+	 * the connection.
+	 * @param remaining The remaining about of data in the packet.
+	 * @param pkt The packet being processed.
+	 */
+	virtual void ContinueProcessing(Connection* conn, double t, bool is_orig, int remaining,
```

Comment:
More generally, the control flow feels hard to follow here: AnalyzerPacket -> ProcessConnection -> ContinueProcessing -> DeliverPacket --- this jumps back and forth between the derived and base class. It would be more easy to follow if one side stayed in charge of the flow and just called the other for help. One idea with the base class as the driver:

1. Implement `AnalyzePacket()` in the base class, have it call a virtual method `BuildConnTuple`, then directly continue with what's in `ProcessConnection`.
2. Once it comes to `ContinueProcessing`, rename that a virtual method `DeliverPacket` that does the dispatching to the transport layer analyzer (and everything else). 
3. Then also pass the current `session_analyzer` around with the method calls, maintaining the member variable seems error prone. 

Alternatively, flip it around: implement `AnalyzerPacket()` inside the derived class and call the base call for help as needed; might make maintaining the `session_analyzer` more natural. not sure.

Reply:
The comments here should all be resolved now. I found a few other things that we didn't need to be storing during processing as well, especially in the ICMP analyzer.